### PR TITLE
sjawhar-legion-427: consolidate webhook receivers into listener

### LINKED ISSUE #427: refactor(envoy): consolidate 3 webhook receiver binaries into single router
## Problem

Three separate webhook receiver binaries on three separate ports:
- Port 9010: GitHub webhook receiver (`cmd/github/main.go`)
- Port 9011: Slack webhook receiver (`cmd/slack/main.go`)
- Port 9012: Ghost Wispr webhook receiver (`cmd/ghostwispr/main.go`)

Each runs as its own Docker container, each has its own compose file, its own Pulumi service function, and its own NATS connection. The listener (`cmd/listener/main.go`, port 9020) is a 4th container. 4 containers × N machines = operational overhead.

All three receivers follow an identical pattern: load config → connect NATS → register HTTP handler → verify signature → parse payload → publish envelope to NATS → return 200/503. The only differences are signature verification scheme and payload normalization.

## Solution

Consolidate all three webhook receivers into the listener process. One binary, one container, one NATS connection.

### Network Topology (per machine)

```
Internet  ──► :9020 (public HTTP)
                ├── /healthz            (always available)
                ├── /webhook/github     (when enabled: HMAC SHA256, gated by NATS readiness)
                ├── /webhook/slack      (when enabled: Slack signing, gated by NATS readiness)
                └── /webhook/ghostwispr (when enabled: optional HMAC, gated by NATS readiness)

Tailscale ──► :443  (tsnet TLS, when ENVOY_TSNET_ENABLED=true)
                └── /v1/*              (internal API, gated by NATS readiness)
```

### Design

**Go code:**

1. Extract webhook handlers from `cmd/{github,slack,ghostwispr}/main.go` into `internal/webhook/{github,slack,ghostwispr}.go`. Each exports a constructor returning `http.HandlerFunc` given config (secrets, triggers) and a publisher interface (not `atomic.Pointer[listenerDeps]` — follow the Ghost Wispr `envelopePublisher` pattern for testability).

2. In `cmd/listener/main.go`, **config-gated activation** for webhook routes:
   - New env var `ENVOY_WEBHOOKS` — comma-separated list of enabled providers (e.g., `github,slack,ghostwispr`). Empty = no webhooks.
   - When a provider is enabled AND its required secret is missing → **startup failure** (preserves fail-fast behavior from current receivers' `getenv()` pattern).
   - When a provider is NOT enabled → route not registered → 404.
   - Ghost Wispr secret remains optional (skip verification when empty), but Ghost Wispr itself must be explicitly enabled via `ENVOY_WEBHOOKS`.

3. Webhook routes are registered on the **public mux** (port 9020), gated by `readinessGate` (return 503 until NATS is connected). This matches the current receiver behavior: current receivers also return errors when NATS is unavailable.

4. Webhook env vars added to listener config:
   - `ENVOY_WEBHOOKS` — comma-separated enabled providers (new)
   - `ENVOY_GITHUB_WEBHOOK_SECRET` — required when github enabled
   - `ENVOY_GITHUB_MENTION_TRIGGER` — optional, default `@legion`
   - `ENVOY_SLACK_SIGNING_SECRET` — required when slack enabled
   - `ENVOY_GHOSTWISPR_SIGNING_SECRET` — optional even when ghostwispr enabled

5. Remove `cmd/github/`, `cmd/slack/`, `cmd/ghostwispr/` directories.

6. The existing `internal/verify/` package stays unchanged.

7. Helper functions currently duplicated across receivers (`getenv`, `addr`, `typeName`, `stringValue`, `writeOK`, ghost wispr normalization functions) are consolidated or inlined.

**Docker:**

1. Dockerfile builds only `envoy-listener`; remove 3 receiver build lines and COPY lines.

**Compose:**

1. Merge `github.compose.yml`, `slack.compose.yml`, `ghostwispr.compose.yml` into `listener.compose.yml`.
2. Add webhook env vars to the listener service (including `ENVOY_WEBHOOKS`).
3. Delete the 3 receiver compose files.

**Pulumi:**

1. Remove `ReceiverConfig` interface and `receivers` field from `MachineConfig`.
2. Add `webhooks?: WebhookConfig` to `ListenerConfig`:
   ```ts
   interface WebhookConfig {
     github?: boolean;
     slack?: boolean;
     ghostwispr?: boolean;
   }
   ```
3. `createListener()`: when `machine.listener.webhooks` is set, compute `ENVOY_WEBHOOKS` env var from enabled flags and pass corresponding secret env vars. Required secrets are always passed when provider is enabled (Pulumi fails at config time if secret is missing).
4. Remove `createGithubReceiver()`, `createSlackReceiver()`, `createGhostWisprReceiver()` from `services.ts`.
5. Remove receiver imports and creation calls from `index.ts`.

**Documentation:**

1. Update `packages/envoy/AGENTS.md` — "Where to look" table and "Receiver behavior" section reflect the new structure (`internal/webhook/` instead of `cmd/*/`).

### Startup and Readiness Semantics

The listener binds port 9020 and serves HTTP **before** NATS is connected (existing behavior — allows `/healthz` to respond during startup). Webhook handlers are behind `readinessGate`, so they return 503 until NATS connects.

This is acceptable because:
- Docker/Pulumi health checks gate traffic routing — the container isn't considered healthy until `/healthz` returns `200 {"status":"healthy"}` (which requires NATS connection)
- The current dedicated receivers also fail on NATS issues (returning 503 when publish fails)
- Webhook providers (GitHub, Slack) retry on 5xx responses

The `/healthz` endpoint behavior:
- Before NATS: `200 {"status":"starting"}` (process alive, not ready)
- After NATS connected: `200 {"status":"healthy"}` (fully operational)
- NATS disconnected: `503 {"status":"unhealthy"}` (degraded)

### Migration / Cutover Plan

1. **Ship new listener binary** with webhook code disabled (no `ENVOY_WEBHOOKS` set) — deploys safely alongside existing receivers
2. **Update Pulumi config**: set `listener.webhooks` on the target machine, add `ENVOY_WEBHOOKS` env var
3. **Deploy**: new listener starts with webhook routes active on port 9020; old receiver containers still running on 9010/9011/9012
4. **Switch webhook URLs** at each provider (GitHub settings, Slack app config, Ghost Wispr settings) from `host:901x/webhook/*` to `host:9020/webhook/*`
5. **Verify** webhook delivery works end-to-end via the new listener
6. **Remove old receivers**: delete receiver compose files, remove Pulumi `receivers` config, remove `cmd/` directories
7. **Rollback**: if issues, revert webhook URLs to `host:901x`; old receivers are still running until explicitly removed

Duplicate delivery risk during overlap is low: envelope dedupe keys are source-derived (`github.<delivery>`, `slack.<event_id>`, `ghostwispr.<delivery>`), so the listener's per-session dedupe suppresses most double-delivery if old and new ingress briefly overlap.

## Acceptance Criteria

- [ ] **Single binary**: Docker image contains only `envoy-listener`; `envoy-github`, `envoy-slack`, `envoy-ghostwispr` binaries are gone
- [ ] **Webhook routing**: `POST /webhook/github`, `POST /webhook/slack`, `POST /webhook/ghostwispr` are handled by the listener process on port 9020 when their provider is enabled via `ENVOY_WEBHOOKS`
- [ ] **Config-gated activation**: Webhook routes are registered only when provider is listed in `ENVOY_WEBHOOKS`. Unlisted provider → 404. Listed provider with missing required secret → startup failure (fail-fast)
- [ ] **Signing verification preserved**: GitHub HMAC SHA256 (`X-Hub-Signature-256`), Slack signing (`X-Slack-Request-Timestamp` + `X-Slack-Signature` with 5-min window), Ghost Wispr optional HMAC (`X-GhostWispr-Signature`) — all work identically to current receivers
- [ ] **Internal API routing**: When `ENVOY_TSNET_ENABLED=true`, `/v1/*` routes served exclusively on tsnet (:443) and absent from public port. When `ENVOY_TSNET_ENABLED=false`, `/v1/*` served on public port (existing fallback behavior, unchanged)
- [ ] **NATS readiness gating**: Webhook handlers return 503 until NATS is connected; `/healthz` returns `200 {"status":"starting"}` pre-NATS, `200 {"status":"healthy"}` post-NATS, `503 {"status":"unhealthy"}` on NATS failure
- [ ] **Behavior parity — GitHub**: Bot sender filtering, @legion mention fan-out to `.comment` + `.mention` topics, `ENVOY_GITHUB_MENTION_TRIGGER` configurability
- [ ] **Behavior parity — Slack**: URL verification challenge response (`url_verification` → echo challenge), `event_callback` type filtering, event_id requirement
- [ ] **Behavior parity — Ghost Wispr**: Only `session_started`/`session_ended`/`summary_ready` published; other events → 200 + log skip; `session_id` required; event type normalization; optional signing
- [ ] **Fail-loud on publish**: Webhook handlers return 503 when NATS publish fails (existing receiver behavior)
- [ ] **Handler testability**: Webhook handlers extracted as functions taking a publisher interface (not listener internals), following the Ghost Wispr `envelopePublisher` pattern
- [ ] **Deployment consolidated**: Single compose file replaces 4; `ReceiverConfig`/`receivers` removed from Pulumi `MachineConfig`; webhook config moved to `ListenerConfig.webhooks`; 3 `create*Receiver` functions removed

## Testing Infrastructure

**Available:**
- Go unit tests (`go test ./...` in `packages/envoy/`)
- Existing `internal/verify/` test coverage
- Ghost Wispr handler already extracted as testable function (`webhookHandler`)
- `bun test` for Pulumi/TypeScript tests

**Gaps:**
- No integration test harness for sending webhook payloads to the listener
- No automated tsnet isolation verification (requires Tailscale network)
- Webhook URL migration (GitHub/Slack/Ghost Wispr external settings) is manual

**Verification approach:**
- Unit tests: All webhook handlers extracted as testable functions with mock publisher interface
- Smoke test: `curl` against local listener for route registration, signature verification, readiness gating, and 503-on-publish-failure
- Config validation: startup with enabled provider + missing secret → process exits non-zero
- Deployment: Pulumi preview shows container count reduction; deploy + verify webhook delivery end-to-end

## Context

- Sami's feedback on #426 (security fix): #426 established the tsnet security boundary; this is the V2 consolidation item
- Related PR: #428 (tsnet config deployment — in progress)
- Security model: webhooks stay public with signing verification; internal API stays tsnet-only
- Learning: `docs/solutions/envoy/contracts-and-handler-patterns.md` — extract handlers for testability
- Relevant skills for downstream workers: `envoy`

### PR DESCRIPTION
**Pull request by sjawhar**
Closes #427

## Summary

Consolidates 3 standalone webhook receiver binaries (GitHub port 9010, Slack port 9011, Ghost Wispr port 9012) into the existing listener process on port 9020. One binary, one container, one NATS connection.

### Changes
- **New `internal/webhook/` package**: Extracted handlers from `cmd/{github,slack,ghostwispr}/main.go` into testable functions behind a narrow `Publisher` interface
- **Config-gated activation**: New `ENVOY_WEBHOOKS` env var (comma-separated) controls which webhook routes are registered. Missing required secret when provider is enabled = startup failure (fail-fast)
- **Webhook routes on public mux**: `/webhook/github`, `/webhook/slack`, `/webhook/ghostwispr` registered behind `readinessGate` (503 until NATS connected)
- **42 new tests**: Handler behavior parity (signing, bot filtering, mention fan-out, url_verification, publish failure), config parsing with edge cases
- **Deployment consolidated**: Single Docker binary, single compose file, Pulumi `ReceiverConfig` → `WebhookConfig` on `ListenerConfig`
- **Deleted**: `cmd/github/`, `cmd/slack/`, `cmd/ghostwispr/`, 3 receiver compose files, 2 receiver deploy scripts, 3 Pulumi `create*Receiver` functions

### CODE REVIEW

**Comment by sjawhar** [comment]
## Behavioral Test Results

### Summary
**PASS** — 12/12 acceptance criteria verified. 1 criterion has a P2 deployment-path caveat (Go-level enforcement correct, Pulumi wiring gap).

### Rebase Note
Branch conflicted on rebase to main (Pulumi.prod.yaml: PR #428 added tsnet config for sami-agents-mx). Resolved by merging both tsnet config + webhook config. All other files applied cleanly.

### Results
| # | Criterion | Status | Evidence |
|---|-----------|--------|----------|
| 1 | Single binary | ✅ | `docker build` succeeds with single `envoy-listener` binary. `packages/envoy/cmd/` contains only `listener/` and `mcp/`. `grep -r "envoy-github\|envoy-slack\|envoy-ghostwispr"` returns clean. |
| 2 | Webhook routing on port 9020 | ✅ | `cmd/listener/main.go` loads `webhook.LoadWebhookConfig()` and conditionally registers `/webhook/github`, `/webhook/slack`, `/webhook/ghostwispr` on public mux. Compose `PORT` defaults to `9020`. |
| 3 | Config-gated activation | ✅ (P2 caveat) | **Go layer correct:** `config.go` errors on empty/whitespace-only GitHub/Slack secrets; unlisted providers not registered → 404. **P2 deployment caveat:** `infra/index.ts` uses `cfg.getSecret()` (optional) and `services.ts` interpolates unconditionally — if Pulumi config is missing the secret, `pulumi.interpolate` resolves `undefined` → string `"undefined"`, bypassing Go's empty-string check. Current prod config has all secrets, so this is theoretical. Fix: guard env push on secret existence in `services.ts`. |
| 4 | Signing verification preserved | ✅ | GitHub: `verify.Github()` (HMAC SHA256 via `X-Hub-Signature-256`). Slack: `verify.Slack()` (timestamp+signature, 5-min window). Ghost Wispr: `verify.GhostWispr()` optional HMAC. All 3 tested with signature computation in `*_test.go`. |
| 5 | Internal API routing (/v1/*) | ✅ | Verified by code inspection: `cmd/listener/main.go` — `/v1/*` on tsnet mux when `tsCfg.Enabled`, on public mux otherwise. No changes to this logic in the PR. |
| 6 | NATS readiness gating | ✅ | Webhook routes wrapped in `readinessGate(func() bool { return deps.Load() != nil }, ...)` — returns 503 before NATS. `/healthz` returns `{"status":"starting"}` pre-NATS, `{"status":"healthy"}` post-NATS, `503 {"status":"unhealthy"}` on failure. |
| 7 | Behavior parity — GitHub | ✅ | Bot filtering: `contracts.GithubIsBotSender()` + `githubEvent()` for issue_comment/PR review types. Mention fan-out: `contracts.GithubEnvelopes()` produces 3 envelopes (base `.comment` + resource `.mention` + repo `.mention`). Configurable trigger via `ENVOY_GITHUB_MENTION_TRIGGER`. Test: `TestGitHubHandler/valid_issue_comment_with_@legion_mention_—_3_envelopes` PASS. |
| 8 | Behavior parity — Slack | ✅ | URL verification before signature check (line 42 `slack.go`). `event_callback` + `event_id` required for publish; other types → 200. Test: `TestSlackHandler/url_verification_echoes_challenge_before_signature_check` PASS. |
| 9 | Behavior parity — Ghost Wispr | ✅ | Only `session_started`/`session_ended`/`summary_ready` published; others → 200 + log skip. `session_id` required. `normalizeGhostWisprEventType()` handles case/dot normalization. Optional signing when secret empty. Tests: 10/10 PASS including `non-publishable_event_type`, `event_type_mismatch`, `missing_session_id`. |
| 10 | Fail-loud on publish | ✅ | All 3 handlers return `503 service unavailable` on `publisher.Publish()` error. Tests: `publish_failure_returns_503` PASS for all handlers. |
| 11 | Handler testability | ✅ | `webhook.go` defines `Publisher` interface + `PublisherFunc` adapter. Handlers extracted as `GitHubHandler()`, `SlackHandler()`, `GhostWisprHandler()` — all take `Publisher`, not listener internals. `mockPublisher` in `webhook_test.go`. All 42 tests run without NATS. |
| 12 | Deployment consolidated | ✅ | Single compose: `listener.compose.yml` (webhook env vars added). `ReceiverConfig`/`receivers` removed from `machines.ts` → `WebhookConfig` on `ListenerConfig`. `index.ts` removed 3 conditional receiver blocks. `services.ts` removed `createGithubReceiver/createSlackReceiver/createGhostWisprReceiver`. `Pulumi.prod.yaml` migrated `receivers` → `listener.webhooks`. No stale references found. |

### Test Quality Critique
- **Strengths:** Table-driven tests across all 4 test files. `mockPublisher` pattern enables testing without NATS. Config tests include edge cases (whitespace-only secrets, unknown providers). Good coverage of critical paths (signing, bot filtering, mention fan-out, URL verification ordering).
- **P2: No envelope content verification.** Tests check HTTP status + publish count but never inspect envelope topics, source, or payload structure. An implementation publishing garbage envelopes with the correct count would pass all tests. Consider adding assertions on `pub.published[0].Topic` and `pub.published[0].Source`.
- **P2: No Slack timestamp window test.** The `verify.Slack()` function enforces a ±5 minute window, but `slack_test.go` always generates fresh timestamps. Add a test with a stale timestamp (>5 min old) to verify rejection.
- **P2: No threaded Slack message test.** `contracts.SlackEnvelopes()` produces an additional envelope when `event.thread_ts` is present, but no test exercises this path.
- **P2: No MaxBytesReader test.** All 3 handlers use `http.MaxBytesReader(w, r.Body, 1<<20)` but no test sends >1MB body to verify 413 response.
- **P2: Missing PR review event test for GitHub.** `githubEvent()` returns true for `pull_request_review_comment` and `pull_request_review`, but only `issue_comment` is tested for bot filtering.

### Documentation Feedback
- `packages/envoy/AGENTS.md` correctly updated: "Where to look" table references `internal/webhook/` instead of old `cmd/*/` paths.
- `deploy/README.md` updated with consolidated structure.
- `deploy/compose/listener.compose.yml` documents all webhook env vars with defaults.
- Setup instructions accurate: `go mod download` → `go build ./cmd/listener/` works cleanly.

### Observations
- **Rebase conflict** with PR #428 (tsnet config) was mechanical — both changes to `Pulumi.prod.yaml` merge cleanly (tsnet under `listener` + webhooks under `listener`).
- **CI workflow** (`.github/workflows/envoy-and-contracts.yaml`) updated to remove old receiver build steps — matches the binary consolidation.

### Config Compliance
- require_specific_task: false — not configured
- require_taiga_evidence: false — not configured

### Build Evidence
```
go build ./cmd/listener/    → exit 0
go test ./internal/webhook/ → 42/42 PASS (0.004s)
go test ./...               → all packages PASS (integration: 138s)
npx tsc --noEmit            → pre-existing deprecation only (moduleResolution=node10)
docker build                → exit 0, single envoy-listener binary
stale reference check       → clean (no envoy-github/slack/ghostwispr refs)
```

**Comment by sjawhar** [comment]
## Review Summary

**CRITICAL (P1):** 0 issues
**IMPORTANT (P2):** 2 issues
**MINOR (P3):** 3 suggestions

**Verdict: Approved to merge.** No critical issues found. All 12 acceptance criteria verified. CI 6/6 green. Two P2 items are real but non-blocking — one is a latent defensive gap in Pulumi (current prod config is safe), the other is test coverage that can be addressed incrementally.

---

### Acceptance Criteria (12/12 ✅)

| # | Criterion | Status | Evidence |
|---|-----------|--------|----------|
| 1 | Single binary | ✅ | Dockerfile builds only `envoy-listener`; `cmd/` contains only `listener/` and `mcp/` |
| 2 | Webhook routing | ✅ | Routes registered on public mux in `main.go:305-322` when provider enabled |
| 3 | Config-gated activation | ✅ | `LoadWebhookConfig()` validates secrets; 12 config_test.go cases cover all edges |
| 4 | Signing verification | ✅ | GitHub: `verify.Github()`, Slack: `verify.Slack()`, Ghost Wispr: `verify.GhostWispr()` |
| 5 | Internal API routing | ✅ | tsnet logic unchanged (`main.go:452-454`); `/v1/*` exclusive to tsnet when enabled |
| 6 | NATS readiness gating | ✅ | Webhook handlers behind `readinessGate`; `/healthz` 3-state behavior preserved |
| 7 | Behavior parity — GitHub | ✅ | Bot filtering, mention fan-out (3 envelopes), sender logging all present |
| 8 | Behavior parity — Slack | ✅ | `url_verification` before signature check; `event_callback` + `event_id` gating; invalid envelope skip |
| 9 | Behavior parity — Ghost Wispr | ✅ | Event type filter (`session_started/ended/summary_ready`); `session_id` required; optional signing |
| 10 | Fail-loud on publish | ✅ | All 3 handlers return 503 on publish failure |
| 11 | Handler testability | ✅ | `Publisher` interface; `mockPublisher` in tests; no NATS dependency |
| 12 | Deployment consolidated | ✅ | Single compose file; `WebhookConfig` on `ListenerConfig`; 3 `create*Receiver` removed |

### Architect/Plan Concerns Verified

All 7 architect concerns and 8 plan concerns are addressed in the implementation. The `envelopePublisher` → `Publisher` pattern was followed correctly. Envelope creation uses `contracts.{GithubEnvelopes,SlackEnvelopes,GhostWisprEnvelope}()` functions — no hand-built envelopes. Non-POST → 200 across all handlers.

### Stale References: Clean

- No references to `envoy-github`, `envoy-slack`, `envoy-ghostwispr` binaries
- No references to `createGithubReceiver`, `createSlackReceiver`, `createGhostWisprReceiver`, `ReceiverConfig`
- No references to deleted compose files (`github.compose.yml`, etc.)
- CI workflows updated (both `envoy-and-contracts.yaml` and `release-envoy-and-plugin.yaml`)

### Injected Learning

`envoy/contracts-and-handler-patterns.md` — handler extraction pattern followed correctly; `Publisher` interface is cleaner than the recommended `*atomic.Pointer[listenerDeps]` approach.

---

### P2 Findings

**P2-1: Pulumi can inject `"undefined"` string for missing webhook secrets** (`infra/services.ts:136-144`)

GitHub and Slack secret env vars are pushed unconditionally when the provider is enabled:
```typescript
if (webhooks.github) {
    webhookEnvs.push(
        pulumi.interpolate`ENVOY_GITHUB_WEBHOOK_SECRET=${secrets.githubWebhookSecret}`
    );
}
```

If `secrets.githubWebhookSecret` is `undefined` (Pulumi config key missing), `pulumi.interpolate` produces `ENVOY_GITHUB_WEBHOOK_SECRET=undefined`. Go config sees non-empty string → passes fail-fast → signature verification silently fails with wrong key.

Ghost Wispr path IS correctly guarded: `if (webhooks.ghostwispr && secrets.ghostWisprSigningSecret)` (line 146). GitHub/Slack should match this pattern.

**Current production impact: None** — all secrets are set in `Pulumi.prod.yaml`.

**P2-2: Test coverage gaps**

Tests verify HTTP status and publish count but have gaps:
- No assertion on envelope topics/source/payload structure
- No Slack timestamp window rejection test (5-min `verify.Slack` window)
- No threaded Slack message test (`thread_ts` → additional envelope)
- No MaxBytesReader (1MB limit) test on any handler
- Missing `pull_request_review` and `pull_request_review_comment` bot filtering tests for GitHub

These don't indicate bugs (handlers are correct), but regressions in envelope content or edge-case behavior can go undetected.

### P3 Suggestions

1. `typeName` and `stringValue` in `slack.go` are identical functions (both do `text, _ := value.(string)`)
2. Inconsistent `MaxBytesError` handling — Ghost Wispr returns 413, GitHub/Slack return 400 for oversized bodies
3. Could unify the `errors` import pattern across handlers

### Evidence
- CI status: All 6/6 checks pass (contracts ✅, envoy-plugin ✅, lint ✅, test ✅, typecheck ✅, envoy-go ✅)
- Go tests: 42 new tests across handler behavior + config parsing
- Stale reference scan: Clean (no orphaned references to deleted binaries/functions/files)
- Tester observations (from test phase handoff): 7 P2 items noted; P2-2 above incorporates them

File: packages/envoy/infra/services.ts [hunk 1]
```diff
@@ -117,11 +117,38 @@ export function createListener(
...
+    const enabled: string[] = [];
+    if (webhooks.github) enabled.push("github");
+    if (webhooks.slack) enabled.push("slack");
+    if (webhooks.ghostwispr) enabled.push("ghostwispr");
+    if (enabled.length > 0) {
+      webhookEnvs.push(`ENVOY_WEBHOOKS=${enabled.join(",")}`);
+    }
+    if (webhooks.github) {
+      webhookEnvs.push(
+        pulumi.interpolate`ENVOY_GITHUB_WEBHOOK_SECRET=${secrets.githubWebhookSecret}`
```
Comment (sjawhar) [inline packages/envoy/infra/services.ts:137]: **[P2]:** `secrets.githubWebhookSecret` can be `undefined` when the Pulumi config key is missing. `pulumi.interpolate` would produce `ENVOY_GITHUB_WEBHOOK_SECRET=undefined` (literal string), which passes Go's non-empty validation but causes all GitHub signature verifications to silently fail with 401.

The ghostwispr path correctly guards against this: `if (webhooks.ghostwispr && secrets.ghostWisprSigningSecret)` (line 146). GitHub and Slack should match.

**Fix:** `if (webhooks.github && secrets.githubWebhookSecret)` — same pattern for Slack on line 141.

**Current prod impact:** None — all secrets set in Pulumi.prod.yaml.

**Comment by sjawhar** [comment]
## Review Summary

**CRITICAL (P1):** 0 issues
**IMPORTANT (P2):** 2 issues
**MINOR (P3):** 3 suggestions

**Verdict: Approved** — No critical issues. All 12 acceptance criteria met. Code correctly consolidates 3 webhook receivers into the listener process with behavior parity preserved.

---

### Acceptance Criteria Checklist

- ✅ **Single binary**: Dockerfile builds only `envoy-listener`; `cmd/github/`, `cmd/slack/`, `cmd/ghostwispr/` deleted
- ✅ **Webhook routing**: `/webhook/{github,slack,ghostwispr}` on port 9020, config-gated via `ENVOY_WEBHOOKS`
- ✅ **Config-gated activation**: `LoadWebhookConfig()` validates required secrets, unknown providers error, empty = no webhooks
- ✅ **Signing verification preserved**: Each handler calls `verify.{Github,Slack,GhostWispr}()` correctly
- ✅ **Internal API routing**: tsnet logic unchanged (`!tsCfg.Enabled` → `/v1/*` on public port)
- ✅ **NATS readiness gating**: Webhooks behind `readinessGate()`; `/healthz` returns starting/healthy/unhealthy
- ✅ **Behavior parity — GitHub**: Bot filtering, mention fan-out (3 envelopes), `ENVOY_GITHUB_MENTION_TRIGGER`
- ✅ **Behavior parity — Slack**: `url_verification` before signing, `event_callback` + `event_id` gate, invalid envelope → continue
- ✅ **Behavior parity — Ghost Wispr**: Only `session_started`/`session_ended`/`summary_ready` published; optional signing
- ✅ **Fail-loud on publish**: All handlers return 503 on NATS publish failure
- ✅ **Handler testability**: All handlers take `Publisher` interface, 42 new tests with `mockPublisher`
- ✅ **Deployment consolidated**: Single compose file, `WebhookConfig` on `ListenerConfig`, 3 `create*Receiver` functions removed

### P2 Findings

**1. Pulumi `undefined` secret injection** (`infra/services.ts:135-144`)
GitHub and Slack secret env vars are pushed unconditionally when the provider is enabled, without checking if the Pulumi secret exists. If `secrets.githubWebhookSecret` is `undefined`, `pulumi.interpolate` produces `ENVOY_GITHUB_WEBHOOK_SECRET=undefined` (literal string), which passes Go's non-empty validation but causes all signature verification to silently fail with 401. The ghostwispr path correctly guards with `&& secrets.ghostWisprSigningSecret` (line 146) — GitHub and Slack should match. Current prod config has all secrets, so this is latent.

**2. Test coverage gaps** (webhook test files)
Tests verify HTTP status codes and publish counts but never inspect envelope topics, source, dedupe_key, or payload structure. Missing coverage: Slack timestamp window rejection, threaded Slack messages (`thread_ts`), MaxBytesReader 1MB limit, GitHub `pull_request_review`/`pull_request_review_comment` bot filtering. These gaps allow regressions in envelope routing semantics without test failure.

### P3 Suggestions

1. `typeName` and `stringValue` in `slack.go` are identical functions — could unify
2. Inconsistent `MaxBytesError` handling: ghostwispr returns 413, github/slack return 400
3. Dead code: `mentionTrigger` empty-string guard in `github.go:93-96` — `LoadWebhookConfig` always provides `@legion` default

### Evidence
- CI status: 6/6 passing (contracts ✓, envoy-plugin ✓, lint ✓, test ✓, typecheck ✓, envoy-go ✓)
- No stale references to old binaries, compose files, or receiver functions
- `cmd/` directory contains only `listener/` and `mcp/`
- `AGENTS.md` correctly updated with new `internal/webhook/` paths
- Tester phase: 12/12 criteria passed, 7 P2 observations (Pulumi undefined secrets + test gaps confirmed by this review)

File: packages/envoy/internal/webhook/webhook_test.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import "github.com/sjawhar/envoy/internal/contracts"
+
+// mockPublisher records published envelopes for test assertions.
+type mockPublisher struct {
+	published []contracts.Envelope
+	err       error
+}
+
+func (m *mockPublisher) Publish(item contracts.Envelope) error {
+	m.published = append(m.published, item)
```
Comment (sjawhar) [inline packages/envoy/internal/webhook/webhook_test.go:12]: **[P2]:** Tests assert on HTTP status and publish count, but never inspect the published envelopes' `.Topic`, `.Source`, `.DedupeKey`, or `.PayloadSummary`. Since `mockPublisher` already records envelopes (webhook_test.go:12), adding assertions like `pub.published[0].Topic` would catch regressions in envelope routing semantics (e.g., wrong topic prefix, missing source field) without requiring NATS.

Additional untested paths:
- Slack 5-min timestamp window rejection
- Slack threaded messages (`thread_ts` → additional envelope)
- MaxBytesReader 1MB body limit (all handlers)
- GitHub `pull_request_review` / `pull_request_review_comment` bot filtering

File: packages/envoy/internal/webhook/slack.go [hunk 3]
```diff
@@ -1,44 +1,29 @@
...
-	if err != nil {
-		log.Fatal(err)
-	}
-	secret := getenv("ENVOY_SLACK_SIGNING_SECRET")
-	client, err := bus.Connect(cfg.NATSURLs, bus.WithReplicas(cfg.NATSReplicas))
-	if err != nil {
-		log.Fatal(err)
-	}
-	defer client.Conn.Close()
+func typeName(value any) string {
```
Comment (sjawhar) [inline packages/envoy/internal/webhook/slack.go:14]: **[P3]:** `typeName` and `stringValue` (lines 14-22) have identical implementations: `text, _ := value.(string); return text`. Could be a single helper. Preserved from original receivers for semantic clarity, so not blocking.
